### ISSUE TITLE
MacOS M1: GLFW Runtime Error

### ISSUE THREAD

**Issue by CyborgPotato** [cloudhead/rx#113]
Cloning master and building with `cargo install --locked --path .` and running the resulting binary results in a blank window and the following error: `GLFW Error: Cocoa: Failed to find service port for display`. Building from the stable branch as on the website using:
```bash
cargo install \
    --git https://github.com/cloudhead/rx \
    --tag v0.4.0
```
Does not result in this error, the only thing I found related to this issue was the following GLFW issue that was fixed earlier this year:
https://github.com/bjornbytes/lovr/pull/357

**Comment by CyborgPotato** [comment]
I have also tried to use the git version of glfw-rs (though it seems to be the same as 0.41), as well as version 0.36, neither of which resolve the runtime glfw error

**Comment by CyborgPotato** [comment]
Fixed in pull request https://github.com/CyborgPotato/rx/tree/M1-GLFW-Fix

**Comment by cloudhead** [comment]
Interesting. Double buffering adds an entire frame of latency, which is not acceptable as a default. Hopefully we can find another solution, or perhaps target only the M1 somehow, or even have a build option to enable double buffering..

**Comment by CyborgPotato** [comment]
It's very interesting, I'll be using the solution for now, but it seems to be an issue with glfw-rs

I'll be investigating glfw's native bindings in C and try to find if double-buffering just isn't supported in GLFW on M1 or if there might be another fix

**Comment by CyborgPotato** [comment]
I have now confirmed that double-buffering causes issue with just GLFW using the following test program:
```c
#include <GLFW/glfw3.h>

int main(void)
{
    GLFWwindow* window;

    /* Initialize the library */
    if (!glfwInit())
        return -1;

    glfwWindowHint(GLFW_DOUBLEBUFFER, GLFW_FALSE);
    
    /* Create a windowed mode window and its OpenGL context */
    window = glfwCreateWindow(640, 480, "Hello World", NULL, NULL);
    if (!window)
    {
        glfwTerminate();
        return -1;
    }

    /* Make the window's context current */
    glfwMakeContextCurrent(window);

    /* Loop until the user closes the window */
    while (!glfwWindowShouldClose(window))
    {
        /* Render here */
        glClear(GL_COLOR_BUFFER_BIT);

        /* Swap front and back buffers */
        glfwSwapBuffers(window);

        /* Poll for and process events */
        glfwPollEvents();
    }

    glfwTerminate();
    return 0;
}
```
It results in the same error as in the rust-glfw version. However changing swapinterval to 0 with `glfwSwapInterval` does not cause this problem.

As such enabling double buffering but setting swap interval to None/0 allows `rx` to run on M1

done with `glfw.set_swap_interval(glfw::SwapInterval::None);`

This also fixes the tearing issue observed.

**Comment by cloudhead** [comment]
Aha, yes this makes sense. Note that `glfwWindowHint(GLFW_DOUBLEBUFFER, GLFW_FALSE);` _disables_ double buffering. So the issue seems to be when double buffering is off and yes the swap interval is left at its default which I guess is "1", ie. vsync. Vsync doesn't really make sense without double buffering, hence the problem. Somehow on Linux there is no issue (in fact setting the swap interval to a different value has no effect) because it seems to just do the right thing.

**Comment by CyborgPotato** [comment]
I apologize, it seems my example code was wrong, in the the GLFW_DOUBLEBUFFER line was commented out, the text proceeding the example describes the fix more aptly:

```It results in the same error as in the rust-glfw version. However changing swapinterval to 0 with glfwSwapInterval does not cause this problem.

As such enabling double buffering but setting swap interval to None/0 allows rx to run on M1

done with glfw.set_swap_interval(glfw::SwapInterval::None);

This also fixes the tearing issue observed.```

**Comment by cloudhead** [comment]
Right, but the issue there is that with double buffering, there is a whole frame of latency (16ms) on Linux. You can check on the M1 by starting rx with `cargo run --release -- --debug`, and checking the frame time in the top corner.

**Comment by CyborgPotato** [comment]
When setting the swapInterval to None/0 that extra 16ms of latency disappears on my system, see the image below:
![image](https://user-images.githubusercontent.com/10701143/117019191-31db7c00-acc3-11eb-940f-6a51eedb0aab.png)
I'll be editing and attaching latency w/ and w/o setInterval None on the M1 as well

**Comment by cloudhead** [comment]
Interesting, so it's only an issue on Linux :thinking:

**Comment by cloudhead** [comment]
> GLFW Buffer True is the same as having buffer true with setInterval with Sync(1),

This doesn't seem right. This means you're not able to get vsync?

**Comment by cloudhead** [comment]
Closed in #115
